### PR TITLE
Add complex support for torch.sum

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38382 Add complex support for torch.sum**

Differential Revision: [D21600127](https://our.internmc.facebook.com/intern/diff/D21600127)

### CODE REVIEW

File: aten/src/THC/THCDeviceUtils.cuh
Comment (zasdfgbnm) [inline aten/src/THC/THCDeviceUtils.cuh:37]: Moved from THC to ATen

File: aten/src/ATen/cuda/DeviceUtils.cuh
Comment (zasdfgbnm) [inline aten/src/ATen/cuda/DeviceUtils.cuh:81]: This is the key to making it work for `c10::complex`.

**Review by anjali411 — APPROVED** [review]
can you also add test for backward [here](https://github.com/pytorch/pytorch/blob/61bea93fca35ad7ace2c9856499d3f478ffb1377/test/test_autograd.py#L4189) by adding `sum` to the white list

**Comment by zasdfgbnm** [comment]
I added sum to the white list. Also, looks like your test in https://github.com/pytorch/pytorch/pull/37959 is more complete. So, after this gets merged, we could rebase #37959 so that this PR will be tested by #37959?

**Comment by anjali411** [comment]
> I added sum to the white list. Also, looks like your test in #37959 is more complete. So, after this gets merged, we could rebase #37959 so that this PR will be tested by #37959?

I think we should add the sum related tests in this PR so that it's fully tested before merge:
```
('sum', 'complex', _small_2d, lambda t, d: [], 1e-2, 1e-2, 1e-5, _complex_types, _cpu_types, False),('sum', 'complex_dim', _small_3d, lambda t, d: [1], 1e-2, 1e-2, 1e-5, _complex_types, _cpu_types, False),
('sum', 'complex_neg_dim', _small_3d, lambda t, d: [-1], 1e-2, 1e-5, 1e-5, _complex_types, _cpu_types, False),
```

**Comment by anjali411** [comment]
great job fixing the rocm bugs @zasdfgbnm  :D

File: test/test_torch.py
Comment (anjali411) [inline test/test_torch.py:17383]: ha! well I added _complex types here https://github.com/pytorch/pytorch/pull/38400/files#diff-9996665f82f52030836eb8657057cfadR17295 but I don't wanna block this PR because of this. let's remove it in the subsequent PR

**Pull request closed without merging**
